### ISSUE TITLE
Please define strategy: [ parallel | maxParallel | matrix ]

### ISSUE THREAD

**Issue by AArnott** [MicrosoftDocs/vsts-docs#1748]
The doc mentions `strategy: [ parallel | maxParallel | matrix ]` but it fails to explain the difference. It seems that `matrix` leads to sequential job scheduling. But what is the difference between `parallel` and `maxParallel`?

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 7458d5b2-1bb7-7d87-7ce6-2b0c430b0155
* Version Independent ID: edf5bbd3-a3f0-5426-e985-31290f8afd05
* Content: [YAML schema - Azure Pipelines](https://docs.microsoft.com/en-us/azure/devops/pipelines/yaml-schema?view=vsts)
* Content Source: [docs/pipelines/yaml-schema.md](https://github.com/MicrosoftDocs/vsts-docs/blob/master/docs/pipelines/yaml-schema.md)
* Product: **devops**
* GitHub Login: @erickson-doug
* Microsoft Alias: **macoope**

**Comment by nschonni** [comment]
Spent a bunch of time trying to figure this out yesterday too. It turns out that it isn't really an "or", but they can be used together https://github.com/MicrosoftDocs/vsts-docs/blob/d46724870ea67672d68add7479aa47bf3bb08c22/docs/pipelines/process/server-phases.md#yaml-2

**Comment by AArnott** [comment]
Thanks. What does `parallel` mean then? Your sample shows using `matrix` and `maxParallel` together.

**Comment by chrisrpatterson** [comment]
parallel is another strategy, you can use that to simply duplicate a job for test distribution scenarios https://docs.microsoft.com/en-us/azure/devops/pipelines/process/phases?view=vsts&tabs=yaml#slicing

**Comment by AArnott** [comment]
Thanks. So it sounds like I can *either* do:

```yml
strategy:
  maxParallel: [number]
  matrix: 
     more stuff
```

OR I can use:

```yml
strategy:
  parallel: [number]
```

Is that right? If so, that would be great to incorporate into the docs instead of showing all three values as siblings, since they are neither all seen together nor always seen separately.

**Comment by chrisrpatterson** [comment]
You can also use maxParallel with parallel.

**Comment by AArnott** [comment]
What would that mean? If I say: 

```xml
strategy:
  parallel: 2
  maxParallel: 4
```

What happens?

**Comment by chrisrpatterson** [comment]
It should just run 2.

**Comment by AArnott** [comment]
So you're saying the two *can* appear together for validation purposes, but really, for documented use cases we actually have just the two I listed above?

**Comment by chrisrpatterson** [comment]
No I am saying there are two job execution strategies in the system, matrix and parallel.  Both of those cause the pipeline to spawn multiple instances of a given job which by default will run in parallel.  There is an additional property that is valid for both strategies that can limit the number of agents that will be used so a single pipeline does not necessarily saturate a pool.

**Comment by AArnott** [comment]
> Both of those cause the pipeline to spawn multiple instances of a given job which by default will run in parallel.

When I used `matrix` without a parallel property, it did _not_ run in parallel. So I don't think parallel is the default.

> There is an additional property that is valid for both strategies that can limit the number of agents that will be used so a single pipeline does not necessarily saturate a pool.

I'm sorry that I'm not getting it. I still don't understand the difference between parallel and maxParallel. You said earlier that if these two properties exist, the `parallel` property (which I had set lower) would be honored and the other would (seem?) to be ignored. So why are there two properties?

**Comment by janpio** [comment]
I can confirm the observation re `matrix` not running anything in parallel by default. I also spent quite some time not trying to understand how to mix and match `matrix`, `parallel` and `maxParallel` - but don't have much success yet :/

**Comment by nschonni** [comment]
I got `matrix` and `maxParallel` to work together. I believe there is an error message about incompatible keys when trying `matrix` and `parallel`

**Comment by janpio** [comment]
I created a related issue https://github.com/MicrosoftDocs/vsts-docs/issues/1770 for some specific language. @PBoraMSFT tries to find someone to help there.

**Comment by janpio** [comment]
`matrix` needed `maxParallel` for me to actually run anything with parallelism. By default it seems to run the `matrix` job sequentally. By setting `maxParallel` to the same number as I have entries in the `matrix`, I could get the wanted behavior.

**Comment by andyjlewis** [comment]
Do we need some fixes to the schema related to #1770 ?

**Comment by vtbassmatt** [comment]
We had a bug where `matrix` definitions didn't parallelize by default. The fix is rolling out. I've already improved the docs as part of #1770.
